enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/LdapRealm.java

```diff
@@ -89,23 +89,29 @@
     private Boolean useSystemAccountForAuthorization;
     private Map<String,Collection<String>> groupToRoleMapping;
     private final SecurityLog securityLog;
+    private final SecureHasher secureHasher;
 
     // Parser regex for group-to-role-mapping
     private static final String KEY_GROUP = "\\s*('(.+)'|\"(.+)\"|(\\S)|(\\S.*\\S))\\s*";
     private static final String VALUE_GROUP = "\\s*(.*)";
     private Pattern keyValuePattern = Pattern.compile( KEY_GROUP + KEY_VALUE_DELIMITER + VALUE_GROUP );
 
-    public LdapRealm( Config config, SecurityLog securityLog )
+    public LdapRealm( Config config, SecurityLog securityLog, SecureHasher secureHasher )
     {
         super();
-        setName( SecuritySettings.LDAP_REALM_NAME );
         this.securityLog = securityLog;
+        this.secureHasher = secureHasher;
+        setName( SecuritySettings.LDAP_REALM_NAME );
         setRolePermissionResolver( PredefinedRolesBuilder.rolePermissionResolver );
         configureRealm( config );
-        setCredentialsMatcher( new HashedCredentialsMatcher( Credential.DIGEST_ALGO ) );
-        setAuthorizationCachingEnabled( true );
-        boolean authenticationCachingEnabled = config.get( SecuritySettings.ldap_authentication_cache_enabled );
-        setAuthenticationCachingEnabled( authenticationCachingEnabled );
+        if ( isAuthenticationCachingEnabled() )
+        {
+            setCredentialsMatcher( secureHasher.getHashedCredentialsMatcher() );
+        }
+        else
+        {
+            setCredentialsMatcher( new AllowAllCredentialsMatcher() );
```

Comment:
No, the LdapRealm fails authentication by throwing an exception if the bind to the ldap server fails, so it is only if caching is enabled that the credentials matcher is needed. But since the credentials matcher is always called it has to be disabled like this (this is actually the default in the JndiLdapRealm super class that this is extending).
The difference now is that previously we always hashed credentials even if it was not being cached, which was just unnecessary work.